Add a `/feed` endpoint that returns an ATOM feed. Fixes #99.

hashhar opened the pull request.

hashhar committed:
Add a `/feed.atom` endpoint for an ATOM feed.

- Add a GET endpoint `/feed.atom` that presents a dynamic Atom 1.0 feed.
- Add media enclosures in feed to make videos available to podcast
  applications.
- Add a link to the feed in `<head>` and on the page footer.
- Use POCO classes for `<feed>`, `<entry>`, `<link>`, `<content> and
  `<author>`.
- Conforms completely to the W3C Feed Validator
  (https://validator.w3.org/feed/).

dnfclas commented:
Hi __@hashhar__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

hishamco commented:
@hashhar I think you missed the calling part to access this from UI

hashhar commented:
@hishamco Oops. I'll add a link. Should I add it just to the Index view or the shared razor files (the header or footer maybe?)

hishamco commented on src/live.asp.net/Controllers/HomeController.cs:
```diff
@@ -37,6 +37,14 @@ public HomeController(IShowsService showsService, ILiveShowDetailsService liveSh
+        [Produces("application/rss+xml")]
+        public async Task<string> GetRssFeed()
+        {
+            var feedService = new FeedService(_showsService, false);
```
Why you set the `disableCache` false all the time?!! if that's your intend I don't think its needed in the constructor, you can passing it directly in `GetRecordedShowsAsync`

hashhar replied:
I was trying to use the same thing as in the `Index` action. I couldn't understand where `disableCache` was coming from.

hishamco replied:
https://github.com/aspnet/live.asp.net/blob/dev/src/live.asp.net/Controllers/HomeController.cs#L24

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -0,0 +1,81 @@
+                // Start iterating over all recorded shows
+                var showsResult = await _showsService.GetRecordedShowsAsync(ClaimsPrincipal.Current, _disableCache);
+                var shows = showsResult.Shows;
+                if (shows == null) return null;
```
Please move the return inside the curly braces

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -0,0 +1,81 @@
+        public async Task<string> GetFeed()
+        {
+            var parent = new StringWriter();
+            var settings = new XmlWriterSettings { Indent = true };
```
Can we simplify producing the feed using POCO and `XmlSerializer`. This is just an idea to clean the code and make it more readable

hashhar replied:
How? Can we pass our own schema to XMLSerializer? Sorry I haven't used XmlSerializer before.

hishamco replied:
My idea is to use POCO, `Rss`, `Channel` and `Item` classes after that we can use the `XmlSerializer` to serialize the `Rss` instance

hashhar replied:
Oh, I understand now. Okay. I'll try implementing and pushing it.

hishamco commented:
IMHO adding the feed links in the footer is the best option 😄

hashhar commented:
I did just that. What about using POCO? I'll try and use it and push a commit when it works.

hishamco commented on src/live.asp.net/Views/Shared/_Layout.cshtml:
```diff
@@ -29,7 +29,7 @@
     <div class="container body-content">
         @RenderBody()
-        
+
```
Nice catch 😄

campersau commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -0,0 +1,84 @@
+                w.WriteEndElement();
+
+                w.WriteElementString("title", "ASP.NET Community Standup Recorded Shows Feed");
+                w.WriteElementString("updated", DateTime.UtcNow.ToString());
```
According to the [Atom spec](https://tools.ietf.org/html/rfc4287) dates should be formatted with [`yyyy-MM-ddTHH:mm:ssK`](https://tools.ietf.org/html/rfc3339)

hashhar replied:
Nice catch. Thanks, I overlooked it.

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -0,0 +1,84 @@
+                w.WriteAttributeString("href", "https://live.asp.net");
+                w.WriteEndElement();
+
+                w.WriteElementString("title", "ASP.NET Community Standup Recorded Shows Feed");
```
```csharp
w.WriteElementString("title", "ASP.NET Community Standup Episodes");
```

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -0,0 +1,84 @@
+        {
+            var parent = new StringWriter();
+            var settings = new XmlWriterSettings { Indent = true };
+            using (var w = XmlWriter.Create(parent, settings))
```
Rename the variable

hishamco commented on src/live.asp.net/Views/Shared/_Layout.cshtml:
```diff
@@ -46,6 +46,7 @@
                             <li><a href="https://www.asp.net/terms-of-use">Terms of Use</a></li>
                             <li><a href="https://github.com/aspnet/live.asp.net">Source Code</a></li>
                             <li><a href="https://github.com/aspnet/live.asp.net/issues">Issue Tracker</a></li>
+                            <li><a href="@Url.Content("~/feed")">Subscribe</a></li>
```
Subscribe is more generic term!!

hashhar commented:
Only the POCO part remains now. Other than that everything works. I also support so that podcast apps can use the videos too.

hishamco commented on src/live.asp.net/Controllers/HomeController.cs:
```diff
@@ -37,12 +38,14 @@ public HomeController(IShowsService showsService, ILiveShowDetailsService liveSh
-        [Produces("application/rss+xml")]
-        public async Task<string> GetRssFeed()
+        [HttpGet("/feed.atom")]
+        [Produces("application/atom+xml")]
```
I was planning to tell you about that 😄

hishamco commented on src/live.asp.net/Views/Shared/_Layout.cshtml:
```diff
@@ -46,7 +47,7 @@
                             <li><a href="https://github.com/aspnet/live.asp.net">Source Code</a></li>
                             <li><a href="https://github.com/aspnet/live.asp.net/issues">Issue Tracker</a></li>
-                            <li><a href="@Url.Content("~/feed")">Subscribe</a></li>
+                            <li><a href="@Url.Content("~/feed")">Subscribe to the Episode Feed</a></li>
```
Same as above

hishamco commented on src/live.asp.net/Views/Shared/_Layout.cshtml:
```diff
@@ -9,6 +9,7 @@
     <title>@ViewData["Title"]</title>
     <meta name="description" content="The weekly ASP.NET Community Standup covers live community Q&amp;A, product team discussion, and the most recent updates. Hosted by Scott Hanselman, Damian Edwards, and Jon Galloway. Recordings of previous meetings are available." />
     <link rel="shortcut icon" href="~/favicon.ico" />
+    <link rel="alternate" type="application/atom+xml" title="ASP.NET Community Standup Episode Feed" href="/feed.atom">
```
Episodes

hishamco commented on src/live.asp.net/Services/YouTubeShowsService.cs:
```diff
@@ -181,7 +181,7 @@ private static class DesignData
                     ThumbnailUrl = "https://img.youtube.com/vi/7O81CAjmOXk/mqdefault.jpg",
                     Url = "https://www.youtube.com/watch?v=7O81CAjmOXk&index=1&list=PL0M0zPgJ3HSftTAAHttA3JQU4vOjXFquF"
-                },
+                }
```
👍

hashhar commented:
@hishamco @campersau This is finished. Nothing more to do here. Should I squash the commits or leave them as is?

Also, the empty default constructors are there only for the `XMLSerializer` to work.

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -18,79 +18,125 @@ public FeedService(IShowsService showsService)
-                writer.WriteElementString("uri", appRoot);
-                writer.WriteEndElement();
+    [XmlRoot("feed")]
+    public class Feed
```
It will be nice if we named `AtomFeed` because `Feed` is a generic term

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -18,79 +18,125 @@ public FeedService(IShowsService showsService)
+    [XmlRoot("feed")]
+    public class Feed
+    {
+        [XmlAttribute("xml:lang")] public string Lang;
```
Use property instead, and move the attribute above

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -18,79 +18,125 @@ public FeedService(IShowsService showsService)
+    [XmlRoot("entry")]
+    public class Entry
+    {
+        [XmlElement("id")] public string Id;
```
Same as previous comment

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -18,79 +18,125 @@ public FeedService(IShowsService showsService)
-                    var thumbnailRenderHtml = string.Format("<a href={0}><img src=\"{1}\"/></a>", show.Url, show.ThumbnailUrl);
-                    writer.WriteString(thumbnailRenderHtml + WebUtility.HtmlEncode(show.Description));
-                    writer.WriteEndElement();
+    [XmlRoot("link")]
```
You can simplify this by adding a property in `Entry` class

hishamco reviewed (changes requested):
Can you please refer to this https://validator.w3.org/feed/docs/atom.html to avoid unnecessary classes

hashhar commented:
Pushed new changes.
- Removed optional elements.
- Uses properties instead of fields.
- Got rid of extra constructors.
- Conforms completely to the W3C Feed Validator now.
- Got rid of xsi and xsd specifications from root element.

hishamco commented on src/live.asp.net/Controllers/HomeController.cs:
```diff
@@ -37,6 +38,16 @@ public HomeController(IShowsService showsService, ILiveShowDetailsService liveSh
+            var feedService = new FeedService(_showsService);
+            var request = Url.ActionContext.HttpContext.Request;
+            var absoluteRoot = new Uri(request.Scheme + "://" + request.Host.Value).ToString();
+            return await feedService.GetFeed(absoluteRoot);
```
Add new line before this as part of code formatting standard

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -0,0 +1,145 @@
+
+        [XmlElement("id")]
+        public string Id { get; set; }
+        [XmlElement("icon")]
```
Please add new line between properties

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    [XmlRoot("entry")]
+    public class Entry
```
Formatting should applied everywhere

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -0,0 +1,145 @@
+            var serializer = new XmlSerializer(typeof(AtomFeed), "http://www.w3.org/2005/Atom");
+            var stringWriter = new StringWriter();
+            serializer.Serialize(stringWriter, feed, feed.Namespaces);
+            return stringWriter.ToString();
```
New line

hishamco reviewed (changes requested):
Can you please refer to this https://blogs.msdn.microsoft.com/dancre/2004/12/13/generating-an-rss-feed-with-xmlserializer-asp-net-and-c/ because I think there 're many POCOs that can replaced by properties

hashhar commented:
@hishamco I think I can't reduce POCOs because any XML Element that acts as a container for other elements needs to have it's own class since the `XmlRoot` annotation can only be applied to classes, strutcs, enums or returns. Something like `<a>` in the example below will need its own class as far as I know.

```
<a>
    <b foo="foo"/>
    <c bar="bar"/>
</a>
```

hishamco commented:
@hashhar did you check the link I posted before?!!

hashhar commented:
@hishamco Yes I read it. The problem is that I have three extra POCOs due to `<author>`, `<content>` and `<link>` tags. This is the minimum due to ATOM specification. But yes, I can get rid of a lot of these annotations by changing the variable names. I didn't know that.

hishamco commented:
I think you're right, my aim to remove unnecessary classes, but after I looked [SyndicationItem](https://referencesource.microsoft.com/#System.ServiceModel/System/ServiceModel/Syndication/SyndicationItem.cs,899eb7ace8d2ea2a) there are Link and Person and Content classes, it would be nice if you check this and come up with a beauty piece of code 😄 
thanks

hashhar commented:
@hishamco Wow, that's so good. I didn't know there was a reference source. I'll definitely take a look. But it may take some time since I have an exam tomorrow and the day after that.

I think this effort can be expanded and later ported over to ASP.NET Core. 😄

hishamco commented:
The reference source is a great source of .NET :), regarding the Syndication Stuff, many of the previous APIs planned to come to the new Core world in upcoming NET STANDARDS

hishamco commented:
FYI if we go with the required elements in the Atom feed we are able to remove unnecessary POCOs, so the remaining will be `AtomFeed` and `Entry`

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -0,0 +1,163 @@
+                    feed.Entry.Add(new Entry
+                    {
+                        Id = show.Url,
+                        Updated = show.ShowDate.DateTime.ToString("u").Replace(' ', 'T'),
```
You can use
```csharp
Updated = show.ShowDate.DateTime.ToString("s") + "Z";
```

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -0,0 +1,163 @@
+        public string Updated { get; set; } = DateTime.UtcNow.ToString("u").Replace(' ', 'T');
+
+        [XmlElement("entry")]
+        public List<Entry> Entry { get; set; }
```
I think `Entries` make sense as property name

hashhar commented:
I simplified it the best I could. I removed the extra annotations where I could because by default XmlElement names are the same as variable names.
I also made the change to DateTime as you suggested and converted Link and Entry to links and entries to make sense because they are Collections.

hishamco commented on src/live.asp.net/Services/FeedService.cs:
```diff
@@ -78,86 +78,73 @@ public AtomFeed()
 
-        [XmlElement("id")]
-        public string Id { get; set; }
+        public string id { get; set; }
```
I think you break the naming conversions for the property again!!

hashhar replied:
Hmmm, properties have to be uppercase only?

hashhar replied:
Because making them lowercase allows me to remove the annotations where not needed. I can simply revert the commit if you want.

JimiC replied:
It's just a standard property naming convention .NET/C# has for years.

hishamco replied:
I know what you gonna say, but always refer to [CONTRIBUTING.md](https://github.com/aspnet/live.asp.net/blob/dev/CONTRIBUTING.md)

BTW what you're doing is good as contributor but you should follow the guidance, I suggest to finish everything and squash the commits if it's possible, after you can wait for **Damian Edwards** approval because he is the man behind the entire stuff in live.asp.net

hashhar replied:
Okay. Thanks. I read the document and will change accordingly. Sorry for not paying attention to it early on.

hishamco commented:
Please have a look to my last blog [here](http://هشام.com/%D8%A5%D9%86%D8%B4%D8%A7%D8%A1-%D8%AA%D9%84%D9%82%D9%8A%D9%85-%D8%A7%D9%84%D9%88%D9%8A%D8%A8-atom-%D9%81%D9%8A-%D8%A7%D9%84%D9%80-asp.net-%D8%A8%D8%A5%D8%B3%D8%AA%D8%AE%D8%AF%D8%A7%D9%85-%D8%A7%D9%84%D9%80-xmlserializer) I know it's an Arabic 😄 , at least you can look to the code snippets, this is the less amount code I come up with to generate an ATOM feed, hope you like it

hashhar commented:
I squashed it all into a single commit. I don't think I can trim it down any more. Changing it to what you have in your blog will break two things.

1. In the current form, the content contains a thumbnail.
2. The media enclosure links allow feed readers to treat this as a podcast and will allow downloading the video.

Should I add doc comments to public functions?

hashhar commented:
I just now tested with my own API key and it works without any issues. See the result.

![Image](http://i.imgur.com/9wVUgO5.png)

hishamco commented:
I didn't mean that you should follow what I did, but what i'm doing is the least amount of code to generate the ATOM feed
I will try to test the code in my end, and let you know if there's any further suggestion, but in general that's great stuff 👍

hashhar committed:
Convert to RC4 (project.csproj)
